src/ser.rs

```diff
@@ -496,11 +492,7 @@ impl<'a> Serializer<'a> {
             State::Array { type_, .. } => type_,
             _ => return Ok(()),
         };
-        if let Some(prev) = prev.get() {
-            if prev != type_ {
-                return Err(Error::ArrayMixedType);
-            }
-        } else {
+        if let None = prev.get() {
             prev.set(Some(type_));
```

Comment:
This may no longer be necessary?

Reply:
Without it some tests (e.g., `simple_show` and `table` in `tests/display.rs`) fail.

Reply:
Hm can you gist the failures? Do you know why this is still needed? If so, can you explain what this field is doing?

Reply:
```
---- simple_show stdout ----
thread 'simple_show' panicked at 'assertion failed: `(left == right)`
  left: `"[1, 2[]"`,
 right: `"[1, 2]"`', test-suite/tests/display.rs:20:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

---- table stdout ----
thread 'table' panicked at 'Unable to represent value as string: ValueAfterTable', src/libcore/result.rs:1165:5


failures:
    simple_show
    table
```

Reply:
Sorry I don't have the time to dig into this right now, but can you try to learn why the tests are failing? We shouldn't have to record type information any more and this may just be related to something else. I doubt that full type information needs to continue to be recorded and it's probably just something like "was anything seen"

Reply:
Sure, I'll look into this :+1:

Reply:
@alexcrichton Yeah so the problem is at: https://github.com/alexcrichton/toml-rs/blob/75129040cd041b380f4f1d8f2685928392aa9f55/src/ser.rs#L989-L1013

It jumps to `None` case and adds additional `[]`. Since without `prev.set(Some(type_));` we don't set a type.

Reply:
Could this be adjusted to a custom enum then? It's basically three states of "not started", "started", and "started as a table"

Reply:
So does that mean that we should basically eliminate the regular types (`integer`, `float`, etc.) in the `enum State` by changing the `&'static str` with a custom `enum` of your described "not started", "started", "started as a table" states? I assume that because we really shouldn't care about the array type at this point since we allow the array to contain multiple types.

Reply:
That sounds good to me!

Reply:
@alexcrichton I've made some changes. I think that we don't even need the `NotStarted` state, since the type is an `Option` within a `Cell`. I might be wrong though. Anyways open for discussion.